Settings app layout

TGBenW opened the pull request:
iPhone Settings app layout:
- table layout, assets, models and extensions
- tap and toggle handling (with console messages)
- custom icons (if not in SF symbols)
- app icon

TGBenW committed:
Basic layout, file structure update

Table added, file structure updated + snapkit added

TGBenW committed:
Settings full layout and functionality

Complete layout and functionality added

TGBenW committed:
Model & design

Model update, custom icons and app icon added

TGBenW committed:
Merge branch 'feature/settings-layout' into develop

Dmitry-Dorodniy commented on iOS12-HW13-Andrey-Manuilov/Sources/ViewControllers/SettingsViewController.swift:
```diff
@@ -0,0 +1,88 @@
+    // MARK: - Variables
+    var models = [[SettingCellModel]]()
+    
+    private func setupModels() {
```
Эту функция обычно помещается прямо в файл с Model
Это не обязанность ViewController данные собирать.

Dmitry-Dorodniy commented on iOS12-HW13-Andrey-Manuilov/Sources/Extensions/SettingsViewController+Extensions.swift:
```diff
@@ -0,0 +1,40 @@
+import UIKit
+
+extension SettingsViewController: UITableViewDataSource {
```
В принципе можно было не выносить в отдельный файл. А оставить его в прямо в SettingsViewController
Именно в этом случае, контроллер у тебя пустой почти. Эти методы особо никак его не перегружают

Dmitry-Dorodniy commented on iOS12-HW13-Andrey-Manuilov/Sources/Extensions/SettingsViewController+Extensions.swift:
```diff
@@ -0,0 +1,40 @@
+        
+        let model = models[indexPath.section][indexPath.row]
+        switch model.accessoryType {
+        case .toggle: break // do nothing for toggle cells
```
Вот это хорошо, что предусмотрел

Dmitry-Dorodniy commented on iOS12-HW13-Andrey-Manuilov/Sources/Models/SettingCellModel.swift:
```diff
@@ -0,0 +1,18 @@
+struct SettingCellModel {
+    let iconName: String
+    let title: String
+    let iconBackgroundColor: UIColor
```
А вот UIKit в модели, это не хорошо.
Вопрос с цветом стоило решить по другому.
Например создать также свой собственный Enum с цветами.
А преобразовывать их уже в конкретный цвет в отдельном методе

Dmitry-Dorodniy commented on iOS12-HW13-Andrey-Manuilov/Sources/ViewControllers/DetailViewController.swift:
```diff
@@ -0,0 +1,18 @@
+    }
+
+    func setDetailText(_ text: String) {
+        detailLabel.text = text // set text of detailLabel
```
Можно было и иконку уже передать для красоты )

Dmitry-Dorodniy commented on iOS12-HW13-Andrey-Manuilov/Sources/Views/SettingTableViewCell.swift:
```diff
@@ -0,0 +1,178 @@
+    }
+    
+    func configure(with model: SettingCellModel) {
+        if let imageFromAssets = UIImage(named: model.iconName) { // load image from assets or SF symbols, fallback to placeholder if not found
```
Тут хорошо разлулил! Молодец

Dmitry-Dorodniy commented on iOS12-HW13-Andrey-Manuilov/Sources/Views/SettingTableViewCell.swift:
```diff
@@ -0,0 +1,178 @@
+        case .toggle(let isOn, let action): // setup toggle switch
+            let toggleSwitch = UISwitch()
+            toggleSwitch.isOn = isOn
+            toggleSwitch.addAction(UIAction(handler: { _ in
```
Интересное решение с замыканием

Dmitry-Dorodniy reviewed (approved):
Отличная работа.
Понравилось, что иконки для самого приложения ставишь.
Творческий подход приятен )

TGBenW committed:
Feedback Fixes

Minor fixes made to fulfil mentors notes:
- icon is sent to detailed view now
- UIColor is not mentioned in the Model anymore
- model setup function is moved out of ViewController

TGBenW committed:
Merge branch 'feature/feedback-fixes' into develop

TGBenW committed:
MVС implementation

TGBenW committed:
Merge branch 'feature/MVC-implementation' into develop

Dmitry-Dorodniy commented on iOS12-HW13-Andrey-Manuilov/Sources/Scenes/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,29 @@
+import UIKit
+
+class DetailViewController: UIViewController {
```
Понятно, что этот экран максимально прост.
Но нужно было идти до конца и реализовывать для этого модуля тоже архитектуру MVC, не добавляя всё во viewDidLoad, вынеся View отдельно.
Ну или как минимум упорядочить здесь всё по методам

Dmitry-Dorodniy commented on iOS12-HW13-Andrey-Manuilov/Sources/Scenes/Controller/Extensions/SettingsViewController+Extensions.swift:
```diff
@@ -0,0 +1,41 @@
+        switch model.accessoryType {
+        case .toggle: break // do nothing for toggle cells
+        default: // create DetailViewController and go
+            let detailVC = DetailViewController()
```
Не очень хорошо, что твой SettingsViewController знает о существовании DetailViewController.
Было бы хорошо сделать переход через координатор или роутер.

Dmitry-Dorodniy commented on iOS12-HW13-Andrey-Manuilov/Sources/Scenes/Controller/Extensions/SettingsViewController+Extensions.swift:
```diff
@@ -0,0 +1,41 @@
+        default: // create DetailViewController and go
+            let detailVC = DetailViewController()
+            let iconImage = UIImage(named: model.iconName) ?? UIImage(systemName: model.iconName) ?? UIImage(systemName: "exclamationmark.triangle.fill")
+            detailVC.setDetailTextAndIcon(model.title, icon: iconImage)
```
Также ты настраиваешь свойства другого вьюконтроллера прямо здесь, это тоже не хорошо.
Можно просто передать туда экземпляр модели и там внутри это всё уже настраивается

Dmitry-Dorodniy commented on iOS12-HW13-Andrey-Manuilov/Sources/Scenes/Controller/SettingsViewController.swift:
```diff
@@ -0,0 +1,43 @@
+    var models = [[SettingCellModel]]()
+    
+    // MARK: - Outlets
+    private lazy var tableView: UITableView = {
```
У тебя тут в принципе вся вью это таблица. Но если мы делаем чисто по MVC, то по хорошему это нужно было вынести в отдельный файл SettingView и там уже настраивать это свойство и констрейнты и таблицу.
А нажатие в делегате таблицы передавать из вью через делегат.

Dmitry-Dorodniy commented on iOS12-HW13-Andrey-Manuilov/Sources/Scenes/Model/SettingCellModel.swift:
```diff
@@ -0,0 +1,18 @@
+import UIKit
```
Тут UIKit не нужен. У тебя нет в модели элементов UI, что хорошо

Dmitry-Dorodniy commented on iOS12-HW13-Andrey-Manuilov/Sources/Scenes/Model/SettingCellModel.swift:
```diff
@@ -0,0 +1,18 @@
+    let iconName: String
+    let title: String
+    let iconBackgroundColor: ModelColor
+    var accessoryType: AccessoryType
```
Наверное стоило назвать этот enum как-то по другому, чтобы он не перекликался прямо с названием от Apple, чтобы не вводить в заблуждение

Dmitry-Dorodniy reviewed (approved):
В целом код хороший. Но немного не дожал до классического понимания MVC.
Будет хорошо, если проведёшь рефакторинг и попробуешь реализовать, согласно комментариям ниже.
